src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -175,11 +175,264 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Bindin
+            {
+                _ = GetWellKnownTypeMember(argIsIndex ? WellKnownMember.System_Index__ctor : WellKnownMember.System_Range__ctor, diagnostics, syntax: syntax);
+                diagnostics.Add(syntax, useSiteInfo);
+                lookupResult.Free();
+                return true;
+            }
+
+            if (TryFindIndexOrRangeIndexerPattern(syntax, lookupResult, receiverOpt: null, receiverType, argIsIndex, out Symbol? patternSymbol, diagnostics, ref useSiteInfo))
+            {
+                info = new IndexerArgumentInfo(patternSymbol);
+                diagnostics.Add(syntax, useSiteInfo);
+                lookupResult.Free();
+                return true;
+            }
+
+            lookupResult.Free();
+            return false;
+        }
+
+        private bool PerformPatternIndexerLookup(
+            SyntaxNode syntax, TypeSymbol receiverType, LookupResult lookupResult,
+            [NotNullWhen(true)] out IndexerArgumentInfo? info, TypeSymbol indexerArgumentType, BindingDiagnosticBag diagnostics, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        {
+            info = null;
+
+            LookupMembersInType(
+                lookupResult,
+                receiverType,
+                WellKnownMemberNames.Indexer,
+                arity: 0,
+                basesBeingResolved: null,
+                LookupOptions.Default,
+                originalBinder: this,
+                diagnose: false,
+                ref useSiteInfo);
+
+            if (lookupResult.IsMultiViable)
+            {
+                var indexerGroup = ArrayBuilder<PropertySymbol>.GetInstance();
+                foreach (Symbol symbol in lookupResult.Symbols)
+                {
+                    Debug.Assert(symbol.IsIndexer());
+                    indexerGroup.Add((PropertySymbol)symbol);
+                }
+
+                BoundExpression indexerArgumentExpr = new BoundIndexOrRangeIndexerPatternValuePlaceholder(syntax, indexerArgumentType);
+                var analyzedArguments = AnalyzedArguments.GetInstance();
+                analyzedArguments.Arguments.Add(indexerArgumentExpr);
+                BoundExpression receiver = new BoundImplicitReceiver(syntax, receiverType);
+                var overloadResolutionResult = OverloadResolutionResult<PropertySymbol>.GetInstance();
+                this.OverloadResolution.PropertyOverloadResolution(indexerGroup, receiver, analyzedArguments, overloadResolutionResult, allowRefOmittedArguments: false, ref useSiteInfo);
+                if (overloadResolutionResult.Succeeded)
+                {
+                    MemberResolutionResult<PropertySymbol> resolutionResult = overloadResolutionResult.ValidResult;
+                    PropertySymbol indexer = resolutionResult.Member;
+                    if (indexer.GetMethod is not null && IsAccessible(indexer, ref useSiteInfo))
```

Comment:
I think we should be checking the accessibility of the getter, not the indexer property.

Here's an analogy with current property pattern: [https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQGYACLEgEQIG8CT7iTYAXEgBQCcB7ABxOpI8OUAG4RmAUxIBzCcwDcJAM5zFAX3okCago3IBhfnU16ALCQCyACgokEASiP5NLkgH0SAXjtMl/dtw8ICTYJGqKAPQRJBIc3Bwk+gDKAAyYAOzYwQAqABZSQryxzACeJFwJsAgSAB6xJADkFAB0nLwNJADGEDAwXKzAUgCuKghMMCTMuVB+nVwwkjUDEt0jUlNSsqwQnZ0SSkoVvuM7ewdQwAA2Esb0OvhqQA=](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQGYACLEgEQIG8CT7iTYAXEgBQCcB7ABxOpI8OUAG4RmAUxIBzCcwDcJAM5zFAX3okCago3IBhfnU16ALCQCyACgokEASiP5NLkgH0SAXjtMl/dtw8ICTYJGqKAPQRJBIc3Bwk+gDKAAyYAOzYwQAqABZSQryxzACeJFwJsAgSAB6xJADkFAB0nLwNJADGEDAwXKzAUgCuKghMMCTMuVB+nVwwkjUDEt0jUlNSsqwQnZ0SSkoVvuM7ewdQwAA2Esb0OvhqQA=)